src/tests/Microsoft.PowerFx.Core.Tests/ValueTests/ValueTests.cs

```diff
@@ -0,0 +1,325 @@
+
+            dynamic d = val.ToObject();
+            Assert.Equal(15.0, d[1].a);
+
+            // Verify runtime json
+            var resultStr = val.Dump();
+
+            Assert.Equal("Table({a:10,str:\"alpha\"},{a:15,str:\"beta\"})", resultStr);
+        }
+                
+        [Fact]
+        public void TableFromRecords()
+        {
+            RecordValue r1 = FormulaValue.NewRecord(new TestRow { a = 10, str = "alpha" });
+            RecordValue r2 = FormulaValue.NewRecord(new TestRow { a = 15, str = "beta" });
+            TableValue val = FormulaValue.NewTable(r1, r2);
+
+            var result1 = ((RecordValue)val.Index(1).Value).GetField("a").ToObject();
+            Assert.Equal(15.0, result1);
+                        
+            dynamic d = val.ToObject();
+            Assert.Equal(10.0, d[0].a);
+
+            // Verify runtime json
+            var resultStr = val.Dump();
+            Assert.Equal("Table({a:10,str:\"alpha\"},{a:15,str:\"beta\"})", resultStr);
+
+            TableValue val2 = NewTableT(r1, r2);
+            Assert.Equal(resultStr, val2.Dump());
+        }
+
+        // Heterogenous table.
+        [Fact]
+        public void TableFromMixedRecords()
+        {
+            var cache = new TypeMarshallerCache();
+            RecordValue r1 = FormulaValue.NewRecord(new { a = 10, b = 20, c = 30 }, cache);
+            RecordValue r2 = FormulaValue.NewRecord(new { a = 11,         c = 31 }, cache);
+            TableValue val = FormulaValue.NewTable(r1, r2);
+
+            // Users first type 
+
+            var result1 = ((RecordValue)val.Index(1).Value).GetField("a").ToObject();
+            Assert.Equal(11.0, result1);
+
+            var result2 = ((RecordValue)val.Index(1).Value).GetField("b");
+            Assert.IsType<BlankValue>(result2);
+            Assert.IsType<NumberType>(result2.Type);        
+        }
+
+        [Fact]
+        public void EmptyTableFromRecords()
+        {
+            // Empty means we can't infer the type from the records passed in. 
+            var records = new RecordValue[0];
+            TableValue val = FormulaValue.NewTable(records);
+
+            Assert.Empty(val.Rows);
+
+            var resultStr = val.Dump();
+            Assert.Equal("Table()", resultStr);
+        }
+
+        // Helper to bypass function overloading and invoke the generic overload. 
+        private static TableValue NewTableT<T>(params T[] rows)
+        {
+            var cache = new TypeMarshallerCache();
+            return FormulaValue.NewTable<T>(cache, rows);
+        }
+
+        // Single Column Table
+        [Fact]
+        public void TableFromPrimitive()
+        {
+            NumberValue r1 = FormulaValue.New(10);
+            NumberValue r2 = FormulaValue.New(20);
+            TableValue val = FormulaValue.NewSingleColumnTable(r1, r2);
+
+            dynamic d = val.ToObject();
+            Assert.Equal(20.0, d[1]); // SCT returned as arrays
```

Comment:
>// SCT returned as arrays [](http://example.com/codeflow?start=38&length=25)

Is it the case for all single-column tables? Even if the schema is not ![Value:?]? For example, `FormulaValue.FromJson("[{\"a\":1},{\"a\":2}]")`

Reply:
Yes.  See TableValue.ToObject()
It's keyed off DType.IsColumn, and then returns an array of the *first* field.

Reply:
That can get confusing - there are reasons to have a single-column record, and we would lose it (including its schema) if we "objectify" it. I don't have a good solution, but I don't like that we change the user's data "silently"

Reply:
Ack. Maybe ToObject() is always returning a table. And we can have a separate ToArray()  that returns as an SCT.